companion-api/backend/file.go

```diff
@@ -0,0 +1,168 @@
+package backend
+
+import (
+	"encoding/csv"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/mcluseau/autorizo/api"
+	"github.com/mcluseau/autorizo/auth"
+)
+
+var yesValues = map[string]bool{
+	"true": true,
+	"yes":  true,
+	"1":    true,
+}
+
+type fileClient struct {
+	filePath string
+}
+
+// New Client to manage users with a csv file backend
+func FileClient(filePath string) Client {
+	return &fileClient{
+		filePath: filePath,
+	}
+}
+
+var _ Client = &fileClient{}
+
+func (fc *fileClient) GetClaims(id string) (claims auth.ExtraClaims, err error) {
+	_, claims, err = fc.getUser(id)
+
+	return
+}
+
+func (fc *fileClient) PutClaims(id string, newClaims auth.ExtraClaims) (err error) {
+
+	passwordHash, _, err := fc.getUser(id)
+	if err != nil {
+		return
+	}
+
+	err = fc.PutUser(id, passwordHash, newClaims)
+	return
+}
+
+func (fc *fileClient) PutUser(id, passwordHash string, claims auth.ExtraClaims) (err error) {
```

Comment:
I disagree with the implementation here. Given that the file backend is for small volumes it's okay to load it in memory, but the reads will fail as the file is written. So it's better to use a `csv.Writer` and do the writes to a new file, pipe-style. Then use `os.Rename` to replace the previous file with the new one. Also, don't forget to use a `sync.Mutex` around all write operations. We could also implement writes as a `chan` of operations with a worker applying the changes, as it allows more optimizations like write-aggregation, but we're not going to expect a lot of performance for this backend so let's not take that complexity for now.

Reply:
2nd try will be better ;)